lib/TPP/Transforms/Bufferize.cpp

```diff
@@ -72,12 +77,51 @@ struct Bufferize : public tpp::impl::BufferizeBase<Bufferize> {
   void runOnOperation() override;
 };
 
+struct DuplicateFill : public tpp::impl::DuplicateFillBase<DuplicateFill> {
+  void runOnOperation() override;
+};
+
+void DuplicateFill::runOnOperation() {
+  IRRewriter rewriter(&getContext());
+
+  (void)getOperation()->walk([&](linalg::FillOp fillOp) {
+    if (!fillOp.hasTensorSemantics())
+      return WalkResult::advance();
+    Value fillVal = fillOp.getResult(0);
+    // We can fold only zero initialization. We duplicate only
+    // if the fill has multiple uses.
+    if (!utils::isZeroTensor(fillVal) || fillOp->hasOneUse())
+      return WalkResult::advance();
+    SetVector<Operation *> forwardSlice;
+    getForwardSlice(fillVal, &forwardSlice);
+    for (size_t idx = /*Skip first user. Use the current fill*/ 1;
+         idx < forwardSlice.size(); idx++)
+      if (auto linalgOp = dyn_cast<linalg::LinalgOp>(forwardSlice[idx]))
+        if (succeeded(linalgx::utils::isContraction(linalgOp))) {
+          assert(linalgOp.getNumDpsInits() == 1);
+          Value outLinalg = linalgOp.getDpsInits()[0];
+          if (outLinalg == fillVal) {
+            rewriter.setInsertionPoint(linalgOp);
+            Operation *clonedOp = rewriter.clone(*fillOp.getOperation());
+            rewriter.replaceUsesWithIf(fillOp->getResults(),
```

Comment:
Don't you also have to check if each user is single user?

Reply:
User of the cloned op? The idea is to clone the operation and replace the uses of the fillOp with the clonedOp only for the current linalg contraction operation.

Reply:
Something like:
```mlir
%buf = tensor.empty()
%fill = linalg.fill ins(%buf)

%0 = linalg.matmul inst(...) outs(%fill)
%1 = linalg.matmul inst(...) outs(%fill)
%add = linalg.add ins(...) outs(%fill)
%2 = linalg.matmul inst(...) outs(%fill)
%sub = linalg.sub ins(...) outs(%fill) 
%3 = linalg.matmul inst(...) outs(%fill)
```
My question is: without this pass, does that code bufferizes to `alloc+copy` for **every** user? Or does it try to look at the lifetime of `%add` and `%sub` or something else?

If the answer is YES, then adding new `empty+fill` patterns is equal or better than the original, because it will "zero" all matmuls (for xsmm) but not the others (add, sub) which they wouldn't anyway. More importantly, if the element-wise ops operate on the matmul result (`%1` chains into `%add` and there are no more users of `%1`), can we not add a `empty+fill` patter to the tail element-wise?

If the answer is NO, then what care must we take to not break bufferization "smartness" when it comes to non-matmul ops (ex. in-place reuse due to liveness analysis)? This is mostly important when the operations do not chain together (ie. they do not kill the previous SSA values).